File: pathform.ts
/**
 * PathForm Parser for TypeScript
 * 
 * Parses PathForm text format into JSON-compatible TypeScript objects.
 */

type PathFormValue = string | number | boolean | null | PathFormObject | PathFormArray;
type PathFormObject = { [key: string]: PathFormValue };
type PathFormArray = PathFormValue[];

interface PathPart {
    type: 'key' | 'index';
    value: string | number;
}

/**
 * Parse PathForm text into a TypeScript object (JSON-compatible).
 * 
 * @param text - PathForm text string
 * @returns Object representing the parsed PathForm structure
 */
export function parsePathform(text: string): PathFormObject {
    const result: PathFormObject = {};
    const lines = text.split('\n');
    
    for (let lineNum = 0; lineNum < lines.length; lineNum++) {
        let line = lines[lineNum].trim();
        
        // Skip blank lines and comments
        if (!line || line.startsWith('#')) {
            continue;
        }
        
        // Parse statement: path = value
        const match = line.match(/^([^=]+?)\s*=\s*(.+)$/);
        if (!match) {
            throw new Error(`Line ${lineNum + 1}: Invalid syntax: ${line}`);
        }
        
        const pathStr = match[1].trim();
        const valueStr = match[2].trim();
        
        // Parse value
        const value = parseValue(valueStr);
        
        // Set path in result
        setPath(result, pathStr, value);
    }
    
    return result;
}

/**
 * Parse a value string into a TypeScript value.
 * 
 * Handles: numbers, booleans, null, strings (bare or quoted)
 * 
 * @param valueStr - Value string to parse
 * @returns Parsed value
 */
export function parseValue(valueStr: string): PathFormValue {
    valueStr = valueStr.trim();
    
    // null
    if (valueStr === 'null') {
        return null;
    }
    
    // booleans
    if (valueStr === 'true') {
        return true;
    }
    if (valueStr === 'false') {
        return false;
    }
    
    // quoted string
    if (valueStr.startsWith('"') && valueStr.endsWith('"')) {
        // Use JSON.parse for proper escaping
        return JSON.parse(valueStr);
    }
    
    // number (try to parse as number)
    // Check if it looks like a number
    const numberMatch = valueStr.match(/^-?\d+(\.\d+)?([eE][+-]?\d+)?$/);
    if (numberMatch) {
        // Parse as float if it has decimal or exponent, otherwise integer
        if (valueStr.includes('.') || /[eE]/.test(valueStr)) {
            return parseFloat(valueStr);
        }
        return parseInt(valueStr, 10);
    }
    
    // bare string (no quotes, no spaces, not a keyword)
    return valueStr;
}

/**
 * Set a value at a given path in the object.
 * 
 * Path format: segment(.segment|\[index\])*
 * Examples: user.name, items[0], config.models[1].name
 * 
 * @param obj - Object to modify
 * @param pathStr - Path string
 * @param value - Value to set
 */
export function setPath(obj: PathFormObject, pathStr: string, value: PathFormValue): void {
    // Parse path into segments and indices
    const parts: PathPart[] = [];
    let current = '';
    let i = 0;
    
    while (i < pathStr.length) {
        if (pathStr[i] === '.') {
            if (current) {
                parts.push({ type: 'key', value: current });
                current = '';
            }
        } else if (pathStr[i] === '[') {
            if (current) {
                parts.push({ type: 'key', value: current });
                current = '';
            }
            // Find closing bracket
            let j = i + 1;
            while (j < pathStr.length && pathStr[j] !== ']') {
                j++;
            }
            if (j >= pathStr.length) {
                throw new Error(`Unclosed bracket in path: ${pathStr}`);
            }
            const indexStr = pathStr.substring(i + 1, j);
            const index = parseInt(indexStr, 10);
            if (isNaN(index) || index < 0) {
                throw new Error(`Invalid array index in path: ${pathStr}`);
            }
            parts.push({ type: 'index', value: index });
            i = j;
        } else {
            current += pathStr[i];
        }
        i++;
    }
    
    if (current) {
        parts.push({ type: 'key', value: current });
    }
    
    if (parts.length === 0) {
        throw new Error(`Empty path: ${pathStr}`);
    }
    
    // Navigate/create structure
    let currentObj: any = obj;
    for (let i = 0; i < parts.length - 1; i++) {
        const part = parts[i];
        const nextPart = parts[i + 1];
        
        if (part.type === 'key') {
            const key = part.value as string;
            if (!(key in currentObj)) {
                // Determine next type based on next part
                if (nextPart.type === 'index') {
                    currentObj[key] = [];
                } else {
                    currentObj[key] = {};
                }
            }
            currentObj = currentObj[key];
        } else if (part.type === 'index') {
            // Ensure it's an array
            if (!Array.isArray(currentObj)) {
                throw new Error(`Path ${pathStr}: Cannot index into non-array`);
            }
            // Extend array if needed
            const index = part.value as number;
            while (currentObj.length <= index) {
                if (nextPart.type === 'index') {
                    currentObj.push([]);
                } else {
                    currentObj.push({});
                }
            }
            currentObj = currentObj[index];
        }
    }
    
    // Set final value
    const finalPart = parts[parts.length - 1];
    if (finalPart.type === 'key') {
        (currentObj as PathFormObject)[finalPart.value as string] = value;
    } else { // index
        if (!Array.isArray(currentObj)) {
            throw new Error(`Path ${pathStr}: Cannot index into non-array`);
        }
        // Extend array if needed
        const index = finalPart.value as number;
        while (currentObj.length <= index) {
            currentObj.push(null);
        }
        currentObj[index] = value;
    }
}

/**
 * Convert PathForm text to JSON string.
 * 
 * @param pathformText - PathForm text string
 * @param indent - JSON indentation (default: 2)
 * @returns JSON string
 */
export function toJSON(pathformText: string, indent: number = 2): string {
    const obj = parsePathform(pathformText);
    return JSON.stringify(obj, null, indent);
}

/**
 * Convert JSON object to PathForm text.
 * 
 * @param jsonObj - JSON object or string
 * @param flat - If true, emit all paths fully qualified (not grouped)
 * @returns PathForm text string
 */
export function fromJSON(jsonObj: PathFormObject | string, flat: boolean = false): string {
    const obj: PathFormObject = typeof jsonObj === 'string' ? JSON.parse(jsonObj) : jsonObj;
    const lines: string[] = [];
    
    function emitPath(prefix: string, value: PathFormValue): void {
        if (typeof value === 'object' && value !== null && !Array.isArray(value)) {
            // Object
            for (const key in value) {
                if (value.hasOwnProperty(key)) {
                    const newPrefix = prefix ? `${prefix}.${key}` : key;
                    emitPath(newPrefix, value[key]);
                }
            }
        } else if (Array.isArray(value)) {
            // Array
            for (let i = 0; i < value.length; i++) {
                const newPrefix = `${prefix}[${i}]`;
                emitPath(newPrefix, value[i]);
            }
        } else {
            // Leaf value
            let valueStr: string;
            if (typeof value === 'string') {
                // Check if needs quotes
                if (value.includes(' ') || /[=#"\\\n\r\t]/.test(value)) {
                    valueStr = JSON.stringify(value);
                } else {
                    valueStr = value;
                }
            } else if (value === null) {
                valueStr = 'null';
            } else if (typeof value === 'boolean') {
                valueStr = value ? 'true' : 'false';
            } else {
                valueStr = String(value);
            }
            
            lines.push(`${prefix} = ${valueStr}`);
        }
    }
    
    emitPath('', obj);
    
    if (!flat) {
        // Sort lines by prefix for better readability
        lines.sort();
    }
    
    return lines.join('\n');
}

// Default export
export default {
    parse: parsePathform,
    parseValue,
    setPath,
    toJSON,
    fromJSON
};

File: pathform.py
"""
PathForm Parser for Python

Parses PathForm text format into JSON-compatible Python objects.
"""

import json
import re
from typing import Any, Dict, Union


def parse_pathform(text: str) -> Dict[str, Any]:
    """
    Parse PathForm text into a Python dictionary (JSON-compatible).
    
    Args:
        text: PathForm text string
        
    Returns:
        Dictionary representing the parsed PathForm structure
    """
    result = {}
    
    for line_num, line in enumerate(text.splitlines(), 1):
        line = line.strip()
        
        # Skip blank lines and comments
        if not line or line.startswith('#'):
            continue
        
        # Parse statement: path = value
        match = re.match(r'^([^=]+?)\s*=\s*(.+)$', line)
        if not match:
            raise ValueError(f"Line {line_num}: Invalid syntax: {line}")
        
        path_str = match.group(1).strip()
        value_str = match.group(2).strip()
        
        # Parse value
        value = parse_value(value_str)
        
        # Set path in result
        set_path(result, path_str, value)
    
    return result


def parse_value(value_str: str) -> Any:
    """
    Parse a value string into a Python value.
    
    Handles: numbers, booleans, null, strings (bare or quoted)
    """
    value_str = value_str.strip()
    
    # null
    if value_str == 'null':
        return None
    
    # booleans
    if value_str == 'true':
        return True
    if value_str == 'false':
        return False
    
    # quoted string
    if value_str.startswith('"') and value_str.endswith('"'):
        # Use JSON decoder for proper escaping
        return json.loads(value_str)
    
    # number (try to parse as number)
    try:
        # Try integer first
        if '.' not in value_str and 'e' not in value_str.lower():
            return int(value_str)
        # Then float
        return float(value_str)
    except ValueError:
        pass
    
    # bare string (no quotes, no spaces, not a keyword)
    return value_str


def set_path(obj: Dict[str, Any], path_str: str, value: Any) -> None:
    """
    Set a value at a given path in the object.
    
    Path format: segment(.segment|\[index\])*
    Examples: user.name, items[0], config.models[1].name
    """
    # Parse path into segments and indices
    parts = []
    current = ''
    i = 0
    
    while i < len(path_str):
        if path_str[i] == '.':
            if current:
                parts.append(('key', current))
                current = ''
        elif path_str[i] == '[':
            if current:
                parts.append(('key', current))
                current = ''
            # Find closing bracket
            j = i + 1
            while j < len(path_str) and path_str[j] != ']':
                j += 1
            if j >= len(path_str):
                raise ValueError(f"Unclosed bracket in path: {path_str}")
            index_str = path_str[i+1:j]
            try:
                index = int(index_str)
                if index < 0:
                    raise ValueError(f"Negative array index in path: {path_str}")
                parts.append(('index', index))
            except ValueError:
                raise ValueError(f"Invalid array index in path: {path_str}")
            i = j
        else:
            current += path_str[i]
        i += 1
    
    if current:
        parts.append(('key', current))
    
    if not parts:
        raise ValueError(f"Empty path: {path_str}")
    
    # Navigate/create structure
    current_obj = obj
    for i, (part_type, part_value) in enumerate(parts[:-1]):
        next_part = parts[i + 1]
        
        if part_type == 'key':
            if part_value not in current_obj:
                # Determine next type based on next part
                if next_part[0] == 'index':
                    current_obj[part_value] = []
                else:
                    current_obj[part_value] = {}
            current_obj = current_obj[part_value]
        elif part_type == 'index':
            # Ensure it's a list
            if not isinstance(current_obj, list):
                raise ValueError(f"Path {path_str}: Cannot index into non-array")
            # Extend array if needed
            while len(current_obj) <= part_value:
                if next_part[0] == 'index':
                    current_obj.append([])
                else:
                    current_obj.append({})
            current_obj = current_obj[part_value]
    
    # Set final value
    final_part_type, final_part_value = parts[-1]
    if final_part_type == 'key':
        current_obj[final_part_value] = value
    else:  # index
        if not isinstance(current_obj, list):
            raise ValueError(f"Path {path_str}: Cannot index into non-array")
        # Extend array if needed
        while len(current_obj) <= final_part_value:
            current_obj.append(None)
        current_obj[final_part_value] = value


def to_json(pathform_text: str, indent: int = 2) -> str:
    """
    Convert PathForm text to JSON string.
    
    Args:
        pathform_text: PathForm text string
        indent: JSON indentation (default: 2)
        
    Returns:
        JSON string
    """
    obj = parse_pathform(pathform_text)
    return json.dumps(obj, indent=indent, ensure_ascii=False)


def from_json(json_obj: Union[str, Dict], flat: bool = False) -> str:
    """
    Convert JSON object to PathForm text.
    
    Args:
        json_obj: JSON string or Python dict
        flat: If True, emit all paths fully qualified (not grouped)
        
    Returns:
        PathForm text string
    """
    if isinstance(json_obj, str):
        obj = json.loads(json_obj)
    else:
        obj = json_obj
    
    lines = []
    
    def emit_path(prefix: str, value: Any) -> None:
        if isinstance(value, dict):
            for key, val in value.items():
                new_prefix = f"{prefix}.{key}" if prefix else key
                emit_path(new_prefix, val)
        elif isinstance(value, list):
            for i, val in enumerate(value):
                new_prefix = f"{prefix}[{i}]"
                emit_path(new_prefix, val)
        else:
            # Leaf value
            if isinstance(value, str):
                # Check if needs quotes
                if ' ' in value or any(c in value for c in '=#"\n\r\t'):
                    value_str = json.dumps(value)
                else:
                    value_str = value
            elif value is None:
                value_str = 'null'
            elif isinstance(value, bool):
                value_str = 'true' if value else 'false'
            else:
                value_str = str(value)
            
            lines.append(f"{prefix} = {value_str}")
    
    emit_path('', obj)
    
    if not flat:
        # Sort lines by prefix for better readability
        lines.sort()
    
    return '\n'.join(lines)


if __name__ == '__main__':
    # Example usage
    example = """# Metadata
task.id = "green_wave_istanbul_001"
task.type = route_plan
task.lang = tr

# Driver / context
driver.id = 987654
driver.experience_level = "intermediate"

city.name = "Istanbul"
city.zone = "Kadikoy"
city.country = "TR"

# Sensor-like inputs
input.current_speed_kmh = 43
input.distance_to_next_light_m = 250
input.next_light_state = green
input.next_light_time_to_change_sec = 18

# Model config
model.name = gpt-5.1
model.temperature = 0.1
model.max_tokens = 256

# Constraints
constraints[0] = "Use friendly, non-technical Turkish"
constraints[1] = "Max 2 short sentences"
constraints[2] = "Recommend safe and legal speeds only"

# Output hints
output.type = object
output.fields[0].name = "advice"
output.fields[0].type = "string"
output.fields[1].name = "recommended_speed_kmh"
output.fields[1].type = "number"
output.fields[2].name = "justification"
output.fields[2].type = "string"
"""
    
    print("Parsing PathForm...")
    result = parse_pathform(example)
    print("\nParsed JSON:")
    print(to_json(example))
    
    print("\n\nRound-trip test:")
    pathform_again = from_json(result)
    print(pathform_again)

